Document how video recordings for real devices work (#3157)
saucelabs/sauce-docs: docs/mobile-apps/features/video-recording.md

---
id: video-recording
title: Video Recording
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Sauce Labs captures a recording of the device screen during your tests. For most test frameworks, this video becomes available upon test completion or as a live stream while the test is in progress.

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

> Once your test job concludes, the recorded video is encoded, downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 30 days.
> 
> A test job will not be marked as finished until the video has been successfully uploaded from your device to the long-term storage. Consequently, very long-running test sessions might experience a more extended post-processing phase, thus increasing the time the device will be busy.

Reviewer: ```suggestion
Once your test job is complete, the recorded video is encoded, downloaded from the device, and uploaded to our long-term storage. Please note that videos are automatically deleted after 30 days.

A test job will only be marked as finished once the video has been successfully uploaded from the device to long-term storage. As a result, lengthy test sessions may face a longer post-processing phase, which can prolong the time the device remains busy.
```

## Live Stream

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

While your test job is active, you can access a live video stream of the device if you have the job's device_session_id. This live stream is accessible via WebSocket at the following URL:

`wss://api.<sauce_region>.saucelabs.com/v1/rdc/socket/alternativeIo/<device_session_id>`

Access to this stream requires Basic Auth.

> Be aware that this is a raw data stream and requires parsing to be viewable. Each message received on this stream represents a screenshot taken on the device in `jpeg` format.

Reviewer: ```suggestion
Please note that this is a raw data stream and must be parsed for viewing. Each message received in this stream corresponds to a screenshot taken on the device in `jpeg` format.
```

---
id: video-recording
title: Video Recording
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Sauce Labs captures a recording of the device screen during your tests. For most test frameworks, this video becomes available upon test completion or as a live stream while the test is in progress.

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

> For testing on real devices, we employ two video recording methods. The primary method involves on-device recording. As a more reliable fallback, we also utilize a low-frame-rate stream of device screenshots, leveraging the built-in screenshot functionality of all devices. While the screenshot-based video offers lower resolution and frame rates compared to on-device recordings, it ensures video availability in a wider range of scenarios.

Reviewer 2: ```suggestion
For testing on real devices, video recording is done in one of two ways:
- On-device recording
  - Provides higher video quality and framerate, but requires an agent app running on the device
- External screenshot-based recording
  - Lower quality and framerate, but requires less set-up on the device
The method of recording is decided automatically at the start of each test, based on device support (with a preference towards on-device recording).
```

It's important to note that a test job will not be marked as finished until the video has been successfully uploaded to long-term storage. Consequently, very long-running test sessions will experience a more extended post-processing phase, thus increasing the overall duration of the job.

## Live Stream

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

While your test job is active, you can access a live video stream of the device if you have the job's device_session_id. This live stream is accessible via WebSocket at the following URL:

`wss://api.<sauce_region>.saucelabs.com/v1/rdc/socket/alternativeIo/<device_session_id>`

Access to this stream requires Basic Auth.

Please be aware that this is a raw data stream and requires parsing to be viewable. The video stream adheres to the open stf standard.

## Known Issues

### Missing or Corrupted Videos

> Occasionally, a job might lack a video in the video_url property, or the provided video file might be corrupted and unplayable.

Reviewer 2: Do we ever use the "job" term in public-facing documentation, or is it usually always "test"?

Author: We are using the terms interchangeably, but when it comes to the test results we are usually talking about 'jobs'. The API itself is also called '/jobs'.

Our goal is to provide a video for 99% of all real device test jobs, and we have internal Service Level Objectives (SLOs) in place to monitor our progress toward this target.

However, we cannot guarantee video availability in every single instance. Unexpected errors can occur during video encoding, or our long-term storage (S3) might experience temporary unavailability.

> In the event of such errors, we refrain from retrying these operations. This is to avoid further delaying the completion time of your job and because we prioritize delivering partial test results over no test results at all.

Reviewer 2: I would rephrase this to sound a bit more positive-focused (i.e. less than 1% failure rate, wow!). In its current form it feels a bit like we're trying to preemptively apologize or find excuses for ourselves.

Also, do we really want to expose the fact that we use S3 for long-term storage? Is the customer persona reading this documentation technical enough to care? If we ever switch providers to GKE this will be out of date.

Author: I thought I sounded positive, haha. Will adjust this.

Author: I rephrased this, let me know what you think.

---
id: video-recording
title: Video Recording
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Sauce Labs captures a recording of the device screen during your tests. For most test frameworks, this video becomes available upon test completion or as a live stream while the test is in progress.

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

For testing on real devices, video recording is done in one of two ways:
- On-device recording
  - Provides higher video quality and framerate, but requires an agent app running on the device
- External screenshot-based recording
  - Lower quality and framerate, but requires less set-up on the device

> The method of recording is decided automatically at the start of each test, based on device support (with a preference towards on-device recording).

Author: Hmm, I think in this case more is more. 

Customers regularly complain about video inconsistencies. I believe we should be upfront with them and try to tell them how things work under the hood, as much as we can. Which I hope results in less complains then.

> The method of recording is decided automatically at the start of each test, based on device support (with a preference towards on-device recording).

Reviewer 2: @mk-ahmed-saucelabs Your suggestion reads more like a marketing blurb than a technical documentation 😛 
If we really don't want this section, I'd remove it completely than reduce it to such a vague sentence. It's only inviting customers to reach out to support asking what "different video recording mechanisms" might mean (which will trickle back down to us).

Otherwise I thought my suggestion was sufficiently abstracted, I've not gone into implementation details, just clarified that there are two fundamentally different methods.

> The method of recording is decided automatically at the start of each test, based on device support (with a preference towards on-device recording).

Reviewer 3: @buxomant I disagree, I believe we just need the customer to know that there's more than one way we do video recording and it's decided depending on the device in order to be able to offer the best quality while making sure we're not compromising reliability.

I would ask the following question for every information I expose in the public facing documentation -> What value does this bring to the customer? If none then it's just creating confusion.
If we believe there's value in educating the customers about the different video-recording methods then let's keep it.

> Once your test job concludes, the recorded video (either from the device or generated from screenshots) is encoded, either directly on the device or in our infrastructure. Subsequently, the encoded video is downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 30 days, or upon explicit deletion of the associated job.

Author: again, the point is to actually communicate technical details and become more 'open source'

> Once your test job concludes, the recorded video (either from the device or generated from screenshots) is encoded, either directly on the device or in our infrastructure. Subsequently, the encoded video is downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 30 days, or upon explicit deletion of the associated job.

Reviewer 2: I tend to agree that the "(either from the device or generated from screenshots)" part is redundant. If we already clarified earlier that there are two recording methods, we don't need to repeat that here. It should be pretty obvious to customers that if they can see a video in the UI, some downloading & uploading must have happened along the way.

But on Ahmed's suggestion, what kind of post-processing do we even do, outside of encoding? IMO the term is too vague (again inviting customers reaching out to support with questions), we should either remove it or properly clarify what it means.

> Once your test job concludes, the recorded video (either from the device or generated from screenshots) is encoded, either directly on the device or in our infrastructure. Subsequently, the encoded video is downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 30 days, or upon explicit deletion of the associated job.

Author: I changed the wording.

A test job will not be marked as finished until the video has been successfully uploaded to long-term storage. Consequently, very long-running test sessions will experience a more extended post-processing phase, thus increasing the overall duration of the job. A long session also makes it more likely that the video will either be missing or corrupted. We therefore recommend to limit the session duration, if the video is important to you.

## Live Stream

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

While your test job is active, you can access a live video stream of the device if you have the job's device_session_id. This live stream is accessible via WebSocket at the following URL:

`wss://api.<sauce_region>.saucelabs.com/v1/rdc/socket/alternativeIo/<device_session_id>`

Access to this stream requires Basic Auth.

> Be aware that this is a raw data stream and requires parsing to be viewable. The video stream adheres to the [OpenSTF](https://github.com/openstf/minicap?tab=readme-ov-file#usage) standard. Each message that you receive on this web socket represents a screenshot taken on the device.

Reviewer 3: I'm not sure it's any beneficial to mention OpenSTF unless it's for search engine optimisation. Also OpenSTF it is not a standard. I believe we might want to specify that it's an mjpeg stream instead.

Author: updated, removed the openstf description, since we are using a modified version of it. 

Especially the way that we are handling device input is custom and is not represented in openSTF.

We will have to specify this in more detail once we are opening up our APIs.

## Limitations

### Missing or Corrupted Videos

> We maintain approximately 99% reliability for video recordings across our real device test jobs. In rare cases, recordings might be missing or corrupted due to unexpected hardware of software failures. Should you experience reproducible systematic issues with missing videos or notice a pattern of video failures, please contact our support team.

Reviewer: ```suggestion
We maintain approximately 99% reliability for video recordings across our real device test jobs. In rare cases, recordings may be missing or corrupted due to unforeseen hardware or software failures. If you encounter reproducible systemic issues with missing videos or observe a pattern of video failures, please contact our support team.
```

> We maintain approximately 99% reliability for video recordings across our real device test jobs. In rare cases, a job might lack a video in the video_url property, or the provided video file might be corrupted. These situations typically occur due to unexpected errors during video encoding or unexpected networking or hardware issues. Should you experience reproducible systematic issues with missing videos or notice a pattern of video failures, please contact our support team.

Reviewer 3: Very well put. Nit picking, Prefer shorter phrases.
```suggestion
We maintain approximately 99% reliability for video recordings across our real device test jobs. In rare cases, recordings might be missing or corrupted due to unexpected hardware of software failures. Should you experience reproducible systematic issues with missing videos or notice a pattern of video failures, please contact our support team.
```

---
id: video-recording
title: Video Recording
---

import useBaseUrl from '@docusaurus/useBaseUrl';

Sauce Labs captures a recording of the device screen during your tests. For most test frameworks, this video becomes available upon test completion or as a live stream while the test is in progress.

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

For testing on real devices, video recording is done in one of two ways:
- On-device recording
  - Provides higher video quality and framerate, but requires an agent app running on the device
- External screenshot-based recording
  - Lower quality and framerate, but requires less set-up on the device
The method of recording is decided automatically at the start of each test, based on device support (with a preference towards on-device recording).

> Once your test job concludes, the recorded video (either from the device or generated from screenshots) is encoded, either directly on the device or in our infrastructure. Subsequently, the encoded video is downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 60 days, or upon explicit deletion of the associated job.

Reviewer 4: AFAIR the assets are being stored only 30 days (https://docs.saucelabs.com/dev/error-messages/#test-assets-have-expired). Are you sure about 60?

Author: ```suggestion
Once your test job concludes, the recorded video (either from the device or generated from screenshots) is encoded, either directly on the device or in our infrastructure. Subsequently, the encoded video is downloaded from the device and uploaded to our long-term storage. Be aware that videos are automatically deleted after 30 days, or upon explicit deletion of the associated job.
```

A test job will not be marked as finished until the video has been successfully uploaded to long-term storage. Consequently, very long-running test sessions will experience a more extended post-processing phase, thus increasing the overall duration of the job. A long session also makes it more likely that there will be encoding issues. We therefore recommend to limit the session duration, if the video is important to you.

## Live Stream

<p><small><span className="sauceGreen">Real Devices Documentation</span></small></p>

While your test job is active, you can access a live video stream of the device if you have the job's device_session_id. This live stream is accessible via WebSocket at the following URL:

`wss://api.<sauce_region>.saucelabs.com/v1/rdc/socket/alternativeIo/<device_session_id>`

Access to this stream requires Basic Auth.

> Be aware that this is a raw data stream and requires parsing to be viewable. The video stream adheres to the [OpenSTF](https://github.com/openstf/stf) standard.

Reviewer 5: Is there a more specific place we can link to? I think without that this websocket is not really usable

Reviewer 2: No idea tbh, and yeah this is an extremely technical point that probably few customers will care about (if they really want to use the websocket, they'll just ask us about the format). Sometimes less is more 😛

Author: I was thinking that more people will care about this with open API coming up. 

I will check if there is more concrete documentation we can point to, I know that @gmegidish recently reimplemented the alternativeIo socket

Author: Ok, I could not find any official documentation besides this: https://github.com/openstf/minicap?tab=readme-ov-file#usage 
We do not seem to adhere to this standard. 

For now we could document that the messages send over the websockets are stream of jpeg images.

Reviewer 5: that link is actually pretty good 🙂

Reviewer 3: Indeed less is more. Please consider removing this.

## Known Issues

### Missing or Corrupted Videos

We maintain approximately 99% reliability for video recordings across our real device test jobs. We have internal Service Level Objectives (SLOs) in place to monitor this metric and continuously make incremental improvements to our video recording infrastructure.

Achieving this level of reliability is particularly challenging given our extensive device fleet, which encompasses a wide range of device models and operating system versions, each with its own unique characteristics and recording capabilities.

Our teams actively monitor video availability and work to optimize the recording processes while balancing reliability with performance considerations.

> In rare cases (usually 1% of all test runs), a job might lack a video in the video_url property, or the provided video file might be corrupted. These situations typically occur due to unexpected errors during video encoding or temporary unavailability in our long-term storage systems. Should you experience reproducible issues with missing videos or notice a pattern of video failures, please contact our support team so we can investigate potential systemic issues.

Reviewer 2: Agree with making it less verbose. Also I'd change the purely technical phrasing of "a job might lack a video in the video_url property" with just what that means conceptually (e.g. the video is missing). We really shouldn't tie up the public documentation with internal implementation details, if that `video_url` field ever changes in the future no one will remember to update this section.

Also also, "systematic" 😛

> In rare cases (usually 1% of all test runs), a job might lack a video in the video_url property, or the provided video file might be corrupted. These situations typically occur due to unexpected errors during video encoding or temporary unavailability in our long-term storage systems. Should you experience reproducible issues with missing videos or notice a pattern of video failures, please contact our support team so we can investigate potential systemic issues.

Author: Right, @buxomant I just wanted to make the distinction between a missing and a corrupted video, but that could be done without mentioning the video_url.

I still believe that we should mention that we have internal monitoring and SLOs for this, so that customers trust us with this number and we actually reduce the amount of service requests for videos.

Ideally I would also like to mention some more explicit reasons for missing or corrupted videos. Are there more concrete things we could expose?

> In rare cases (usually 1% of all test runs), a job might lack a video in the video_url property, or the provided video file might be corrupted. These situations typically occur due to unexpected errors during video encoding or temporary unavailability in our long-term storage systems. Should you experience reproducible issues with missing videos or notice a pattern of video failures, please contact our support team so we can investigate potential systemic issues.

Author: thank you, applied your wording, much clearer now.
